tensorflow/contrib/go/tensor.go

```diff
@@ -0,0 +1,763 @@
+	return fmt.Sprintf("The specified index %d is out of the dimension  %d range: %d", e.index, e.dim, e.dimsRange)
+}
+
+// ErrSliceExpected The argument must be an Slice.
+type ErrSliceExpected struct {
+	dataType string
+}
+
+func (e *ErrSliceExpected) Error() string {
+	return fmt.Sprintf("The argument must be an Slice, but the data type is: '%s'", e.dataType)
+}
+
+// ErrDataTypeNotSupported The data type is still not suported.
+type ErrDataTypeNotSupported struct {
+	dataType string
+}
+
+func (e *ErrDataTypeNotSupported) Error() string {
+	return fmt.Sprintf("The type of the provided data is still not suported: '%s'", e.dataType)
+}
+
+func serialize(data interface{}, deep int, dimsIn [][]int64) (ser []interface{}, dims [][]int64, dataType DataType, dataSize int64, err error) {
+	v := reflect.ValueOf(data)
+	dims = dimsIn
+
+	if len(dims) == deep {
+		dims = append(dims, []int64{int64(v.Len())})
+	}
+	// Check the value of the elements on this slice, if they are still
+	// slices call to recursivity, if now just add the results
+	switch v.Type().Elem().Kind() {
+	case reflect.Slice:
+		for i := 0; i < v.Len(); i++ {
+			var intSer []interface{}
+			intSer, dims, dataType, dataSize, err = serialize(v.Index(i).Interface(), deep+1, dims)
+			if err != nil {
+				return
+			}
+			ser = append(ser, intSer...)
+		}
+	default:
+		dataSize = int64(v.Type().Elem().Size())
+		dataType, err = getDataTypeFromReflect(v.Type().Elem().Kind(), dataSize)
+		if err != nil {
+			return
+		}
+		for i := 0; i < v.Len(); i++ {
+			ser = append(ser, v.Index(i).Interface())
+		}
+	}
+
+	return
+}
+
+func getDataTypeFromReflect(refType reflect.Kind, dataSize int64) (dataType DataType, err error) {
+	switch refType {
+	case reflect.Int:
+		if cBytesInt32 == dataSize {
+			dataType = DtInt32
+		} else {
+			dataType = DtInt64
+		}
+	case reflect.Int8:
+		dataType = DtInt8
+	case reflect.Int16:
+		dataType = DtInt16
+	case reflect.Int32:
+		dataType = DtInt32
+	case reflect.Int64:
+		dataType = DtInt64
+	case reflect.Uint8:
+		dataType = DtUint8
+	case reflect.Uint16:
+		dataType = DtUint16
+	case reflect.Float32:
+		dataType = DtFloat
+	case reflect.Float64:
+		dataType = DtDouble
+	case reflect.String:
+		dataType = DtString
+	default:
+		err = &ErrDataTypeNotSupported{
+			dataType: refType.String(),
+		}
+		return
+	}
+
+	return
+}
+
+func newTensor(dataType DataType, shape TensorShape, data unsafe.Pointer, size int64) (*Tensor, error) {
+	var dims *int64
+	var llDims []C.longlong
+	var tensorShape *pb.TensorShapeProto
+
+	// Move the data to C allocated memory
+	shapes := 0
+	for _, v := range shape {
+		shapes += len(v)
+	}
+	if len(shape) != 0 {
+		tensorShape = &pb.TensorShapeProto{
+			Dim: make([]*pb.TensorShapeProto_Dim, len(shape)),
+		}
+		llDims = make([]C.longlong, shapes)
+		for i, v := range shape {
+			tensorShape.Dim[i] = &pb.TensorShapeProto_Dim{
+				Size: v[0],
+			}
+
+			for _, s := range v {
+				llDims[i] = C.longlong(s)
```

Comment:
I'm confused - is this a bug?  The code iterates over v, and *overwrites* the previous assignment to llDims[i] with each new result.  (i is invariant inside this inner range loops).  If the goal is only to apply the *last* element of v, then it'd be more clear to use len(v)-1 directly.